In [ ]:
# Files
fhand = open('mbox.txt')
print(fhand)

In [ ]:
# new line character
stuff = 'hello\nworld'
print(stuff,len(stuff))

In [ ]:
# to get the number of lines in the file
fhand = open('mbox.txt')
count = 0
for line in fhand:
    count = count + 1
print('Line Count:',count)

In [ ]:
# smaller files can be read into one string using read

fhand = open('mbox.txt')
inp = fhand.read()
print((type(inp)))

In [ ]:
# print only first 20 characters
print(inp[:20])

In [ ]:
# each call to read exhaust the resource
print(len(fhand.read()))

In [ ]:
# print only lines that starts with prefix From:

fhand = open('mbox.txt')
count = 0
for line in fhand:
    if line.startswith('From:'):
        print(line)
        count = count + 1
print(count)

In [ ]:
# using continue to skip uninteresting lines
fhand = open('mbox.txt')
for line in fhand:
    line = line.rstrip()
    # skip our uninterestingline
    if not line.startswith('From:'):
        continue
    # process our interesting line
    print(line)

In [ ]:
# find a string in the text file

fhand = open('mbox.txt')
for line in fhand:
    line = line.strip()
    # if not found
    if line.find('@uct.ac.za') == -1:
        continue
    # if found
    print(line)


In [ ]:
# letting the user choose the filename and get the count of number of subject lines
fname = input('Enter the file name')
fhand = open(fname)
count = 0
for line in fhand:
    line = line.strip()
    if line.startswith('Subject:'):
        count = count + 1
print(
'There were ',count,'subject lines in the text')

In [ ]:
# using try,except and open
fname = input('Enter the file name')
try:
    fhand = open(fname)
except: 
    print('File cannot be opened:',fname)
    exit()
    
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count = count + 1
print('Number of lines starting with Subject: ',count)

In [ ]:
# writing files
# if file already existing it clears the old data and start fresh
fout = open('output.txt','w')
print(fout)

In [ ]:
# write method does not add new line automatically
line1 = 'This here is the wattle,\n'
line2 = 'the emblem of our land,\n'
fout.write(line1)
fout.write(line2)

In [ ]:
fout.close()

In [ ]:
s = '1 2\t 3\n 4'
print(s)

In [ ]:
# to view whitespace
print(repr(s))

### Exercises
#### Exercise 1
Write a program to read through a file  and print the contents of the file all in upper case.


In [ ]:
fhand = open('mbox-short.txt')

In [ ]:
for line in fhand:
    line = line.strip()
    line = line.upper()
    print(line)


#### Exercise 2
When you encounter a line that starts with “X-DSPAM-Confidence:”
pull apart the line to extract the floating-point number on the line.
Count these lines and then compute the total of the spam confidence
values from these lines. When you reach the end of the file, print out
the average spam confidence.

In [ ]:
count = 0
total = 0
fname = input('Enter the filename')
fhand = open(fname)
for line in fhand:
    line = line.strip()
    if not line.startswith('X-DSPAM-Confidence'):
        continue
    line = line[-6:]
    line = float(line)
    count = count +1
    total = total + line
average = total/count
print(average)

#### Exercise 3
Sometimes when programmers get bored or want to have a
bit of fun, they add a harmless Easter Egg to their program. Modify
the program that prompts the user for the file name so that it prints a
funny message when the user types in the exact file name “na na boo
boo”. The program should behave normally for all other files which
exist and don’t exist. 

In [ ]:

fname = input('Enter the filename')
if fname == 'na na boo boo':
    print("NA NA BOO BOO TO YOU - You have been punk'd!")
try:
    fhand = open(fname)
except:
    print('File {} cannot be opened'.format(fname))
    exit()

count = 0
total = 0
    
for line in fhand:
    line = line.strip()
    if not line.startswith('X-DSPAM-Confidence'):
        continue
    line = line[-6:]
    line = float(line)
    count = count +1
    total = total + line
average = total/count
print(average)

## Lists
### A list is a sequence


In [ ]:
cheeses = ['cheddar','Edam','Gouda']
numbers = [17,123]
empty = []
print(cheeses, numbers, empty)

### Lists are mutable

In [ ]:
print(cheeses[0])

In [ ]:
numbers[1] = 5
numbers

In [ ]:
'Edam' in cheeses

### Traversing a list

In [ ]:
for cheese in cheeses:
    print(cheese)

In [ ]:
# to wrtie to the list
for i in range(len(numbers)):
    numbers[i] = numbers[i]*2
    print(numbers[i])

In [ ]:
for x in empty:
    print('this never happens')

In [ ]:
neslst = ['spam',1, ['Brie','Roquefort','Pol leVeq'],[1,2,3]]
print(len(neslst))

### List operations

* '+' operator concatenates list
* '*' repeats the list a given number of times


In [ ]:
a = [1,2,3]
b = [4,5,6]
c = a + b
print(c)

In [ ]:
[0] * 4

### List slices


### List  methods

### Lists and functions

In [ ]:
# compute the average of list of numbers entered by the user using a list

# without list
total = 0
count = 0
while(True):
    inp = input('Enter a number: ')
    if inp == 'done': break
    value = float(inp)
    total = total + value
    count = count + 1
average = total/count
print(average)

In [ ]:
# using list
numlist = list()
while True:
    inp = input('Enter the number: ')
    if inp == 'done': break
    value = float(inp)
    numlist.append(value)
print(len(numlist))
average = sum(numlist) / len(numlist)
print('Average: ',average)

In [ ]:
n = [3,4,4,5,3,3]

In [ ]:
sum(n)

In [ ]:
# Lists and strings

s = 'spam'
t = list(s)
t

In [ ]:
s = 'pinnning for the fjords'
t = s.split()
t

In [ ]:
s = 'spam-spam-spam'
delimiter = '-'
s.split(delimiter)


In [ ]:
# join is the inverse of split
t = ['pining','for','the','fjords']
delimiter = ' '
delimiter.join(t)

In [ ]:
# Parsing lines

fhand = open('mbox-short.txt')
for line in fhand:
    if not line.startswith('From'):
        continue
    print(line)
    wlist = line.split()
    if len(wlist) == 2:
        continue
    words = line.split()
    print(words[2])


In [ ]:
# Objects and values

a = 'banana'
b = 'banana'
a is b

In [ ]:
# Python created only one string object and both a & b refer to it

a = [1,2,3]
b = [1,2,3]
a is b

In [ ]:
# Two lists are equivalent but are not identical

# Aliasing

a = [1,2,3]
b = a
a is b

In [ ]:
# the association of a variable with an object is called reference

# object with more than one reference is said to be aliased

# changes made with one alias affect the other

b[0] = 17
a

In [ ]:
# for string it never makes a difference whether a & b refer to the same string or not

List Arguments


In [ ]:
# A reference of the list is passed to the function
def delete_head(t):
    del t[0]
    
letter = ['a','b','c','d']
letter


In [ ]:
delete_head(letter)

In [ ]:
letter

In [ ]:
# append method modifies the list, + operator creates a new list

t1 = [1,2]
t2 = t1.append(3)
t1

In [ ]:
print(t2)

In [ ]:
t3 = t1 + [3]
t3

In [ ]:
t3

In [ ]:
t2 is t3

In [ ]:
# below function does not delete the head of a list

def bad_delete_head(t):
    t = t[1:]  # wrong
    
def tail(t):
    return t[1:]

letters = ['a','b','c']
rest = tail(letters)
print(rest)


### Exercis 1
WWrite a function called chop that takes a list and modifies
it, removing the first and last elements, and returns None. Then write
a function called middle that takes a list and returns a new list that
contains all but the first and last elements.

In [ ]:
# writing a function called chop

def chop(t):
    t = t[1:]
    t = t[:-2]

def middle(t):
    t = t[1:-1]
    return t

letters = ['a','b','c','d']
# letters = middle(letters)
# print(letters)
chop(letters)
print(letters)

## Debugging

#### Most list methods modify the argument  and return None. This is the opposite of string methods which return a new string leave the orgiginal alone

In [ ]:
t = t.sort() # WRONG! returns None

In [ ]:
t = ['fjords','for','pining','the']

In [ ]:
t.append('niju')

In [ ]:
x = 5
t.append(x)
t = t + [x]

#Wrong:
t.append([x])
t = t.append(x)
t + x
t = t + x

In [ ]:
t = ['fjords','for','pining','the']
tsrt = sorted(t)
tsrt

In [ ]:
# parsing lines using list

fhand = open('mbox.txt')
for line in fhand:

    words = line.split()
    print('debug:',words)
    if words[0] != 'From':
        continue
    print(words)
    #print(words[2])

In [3]:
# after handling error
fhand = open('mbox.txt')
for line in fhand:

    words = line.split()
    #print('debug:',words)
    if len(words) == 0:
        continue
    if words[0] != 'From':
        continue
    print(words[2])

Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Tue
Tue
Mon
Mon
Mon
Mon
Mon
Mon
Mon
Sun
Sun
Sun
Sun
Sun
Sun
Sun
Sat
Sat
Sat
Sat
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Wed
Wed
Wed
Tue
Mon
Mon
Sun
Sat
Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Wed
Wed
Wed
Wed
Wed
Wed
Wed


### Exercise 2

Figure out which line of the above program is still not
properly guarded. See if you can construct a text file which causes the
program to fail and then modify the program so that the line is properly
guarded and test it to make sure it handles your new text file.



In [ ]:


# Get the input file name

fname = input('Enter the file name: ')

# create file handler objec
fhand = open(fname)

# initializing counter for filtering rows
counter = 0
for line in fhand:

    line = line.strip()
    words = line.split()
    if len(words) == 0:
        continue
    if not words[0] == 'From':
        continue
    print(words[2])
    counter += 1

    if counter == 10:
        break




### Exercise 3
Rewrite the guardian code in the above example without two if statements. Instead,use a compound logical expression using the and logical operator with a single if statement

In [2]:


# Get the input file name

fname = input('Enter the file name: ')

# create file handler object
try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()
# initializing counter for filtering rows
counter = 0
for line in fhand:

    line = line.strip()
    words = line.split()
    if (len(words) == 0 or len(words) == 2) or not words[0] == 'From':
        continue
    # if not words[0] == 'From':
    #     continue
    print(words[2])
    counter += 1

    if counter == 10:
        break




Enter the file name: mbox-short.txt
Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri


#### Exercise 4

Download a copy of Romeo.txt. Write a program to open the file romeo.txt and read it line by line. For each line, split the line into a list of words using the split function. For each word check to see if the word is already in a list. if the word is not in the list add it to the list. When the program completes, sort and print the resulting words in alphabetical order.


In [1]:
fhand = open('romeo.txt')

word_lst = []

for line in fhand:
    words = line.split()
    for word in words:
        if word not in word_lst:
            word_lst.append(word)
word_lst = sorted(word_lst)
print(word_lst)




['Arise', 'But', 'It', 'Juliet', 'Who', 'already', 'and', 'breaks', 'east', 'envious', 'fair', 'grief', 'is', 'kill', 'light', 'moon', 'pale', 'sick', 'soft', 'sun', 'the', 'through', 'what', 'window', 'with', 'yonder']


### Exercise 5

Write a program to read through the mail box data and when you find line that starts with "From". you will split the line into words using the split function. We are interested in who sent the message, which is the second word on the From line.You will parse the from line and print out the second word for each from line. then you will also count the number of from lines and print out a count at the end. 

In [3]:
fhand = open('mbox.txt')

counter = 0
for line in fhand:
    words = line.split()

    if len(words) == 0 or not words[0] == 'From':
        continue
    print(words[1])
    counter += 1
print('Total Count:', counter)


stephen.marquard@uct.ac.za
louis@media.berkeley.edu
zqian@umich.edu
rjlowe@iupui.edu
zqian@umich.edu
rjlowe@iupui.edu
cwen@iupui.edu
cwen@iupui.edu
gsilver@umich.edu
gsilver@umich.edu
zqian@umich.edu
gsilver@umich.edu
wagnermr@iupui.edu
zqian@umich.edu
antranig@caret.cam.ac.uk
gopal.ramasammycook@gmail.com
david.horwitz@uct.ac.za
david.horwitz@uct.ac.za
david.horwitz@uct.ac.za
david.horwitz@uct.ac.za
stephen.marquard@uct.ac.za
louis@media.berkeley.edu
louis@media.berkeley.edu
ray@media.berkeley.edu
cwen@iupui.edu
cwen@iupui.edu
cwen@iupui.edu
ray@media.berkeley.edu
cwen@iupui.edu
zqian@umich.edu
cwen@iupui.edu
zqian@umich.edu
zqian@umich.edu
zqian@umich.edu
mmmay@indiana.edu
cwen@iupui.edu
zqian@umich.edu
zqian@umich.edu
zqian@umich.edu
cwen@iupui.edu
zqian@umich.edu
cwen@iupui.edu
ray@media.berkeley.edu
zqian@umich.edu
zqian@umich.edu
zqian@umich.edu
cwen@iupui.edu
zqian@umich.edu
gopal.ramasammycook@gmail.com
gsilver@umich.edu
gsilver@umich.edu
stuart.freeman@et.gatech.edu
zqian@umic

zqian@umich.edu
dlhaines@umich.edu
gopal.ramasammycook@gmail.com
jimeng@umich.edu
josrodri@iupui.edu
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
nuno@ufp.pt
wagnermr@iupui.edu
cwen@iupui.edu
ian@caret.cam.ac.uk
ian@caret.cam.ac.uk
gopal.ramasammycook@gmail.com
stuart.freeman@et.gatech.edu
cwen@iupui.edu
ian@caret.cam.ac.uk
ian@caret.cam.ac.uk
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
mmmay@indiana.edu
gjthomas@iupui.edu
gjthomas@iupui.edu
dlhaines@umich.edu
dlhaines@umich.edu
chmaurer@iupui.edu
zqian@umich.edu
jimeng@umich.edu
gjthomas@iupui.edu
gjthomas@iupui.edu
jimeng@umich.edu
zqian@umich.edu
cwen@iupui.edu
arwhyte@umich.edu
chmaurer@iupui.edu
cwen@iupui.edu
chmaurer@iupu

zqian@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
john.ellis@rsmart.com
ajpoland@iupui.edu
chmaurer@iupui.edu
antranig@caret.cam.ac.uk
antranig@caret.cam.ac.uk
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
zqian@umich.edu
antranig@caret.cam.ac.uk
zqian@umich.edu
rjlowe@iupui.edu
zqian@umich.edu
zqian@umich.edu
rjlowe@iupui.edu
gopal.ramasammycook@gmail.com
zqian@umich.edu
kimsooil@bu.edu
kimsooil@bu.edu
kimsooil@bu.edu
zqian@umich.edu
ostermmg@whitman.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
zqian@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
jimeng@umich.edu
ray@media.berkeley.edu
ostermmg@whitman.edu
josrodri@iupui.edu
ostermmg@whitman.edu
ostermmg@w

rjlowe@iupui.edu
hu2@iupui.edu
dlhaines@umich.edu
a.fish@lancaster.ac.uk
a.fish@lancaster.ac.uk
a.fish@lancaster.ac.uk
gsilver@umich.edu
mbreuker@loi.nl
mbreuker@loi.nl
mbreuker@loi.nl
mbreuker@loi.nl
mbreuker@loi.nl
mbreuker@loi.nl
ajpoland@iupui.edu
rjlowe@iupui.edu
rjlowe@iupui.edu
a.fish@lancaster.ac.uk
a.fish@lancaster.ac.uk
a.fish@lancaster.ac.uk
dlhaines@umich.edu
mbreuker@loi.nl
david.horwitz@uct.ac.za
zach.thomas@txstate.edu
jimeng@umich.edu
zqian@umich.edu
jzaremba@unicon.net
colin.clark@utoronto.ca
ray@media.berkeley.edu
ray@media.berkeley.edu
ray@media.berkeley.edu
ray@media.berkeley.edu
louis@media.berkeley.edu
dlhaines@umich.edu
dlhaines@umich.edu
dlhaines@umich.edu
dlhaines@umich.edu
ajpoland@iupui.edu
wagnermr@iupui.edu
zqian@umich.edu
ktsao@stanford.edu
ray@media.berkeley.edu
aaronz@vt.edu
zqian@umich.edu
aaronz@vt.edu
bahollad@indiana.edu
aaronz@vt.edu
zqian@umich.edu
ray@media.berkeley.edu
zqian@umich.edu
zqian@umich.edu
zqian@umich.edu
bahollad@indiana.edu
ray@media

### Exercise 6
Rewrite the program that prompts the user for a list of numbers and prints out the maximum and minimum of the numbers at the end when the user
enters done. Write the program to store the numbers the user enters in a list and use the max() and min() function to compute the maximum 
and minimum numbers after the loop completes.

In [6]:
num_lst = []

while True:
    inp = input('Enter numbers: ')
    if inp == 'done':
        break
    elif inp == '' or inp in num_lst:
        print('Enter a number or the number is already existing')
    else:
        num_lst.append(float(inp))

print(num_lst)
print('The maximum number is: ', max(num_lst))
print('The minimum number is: ', min(num_lst))

Enter numbers: 34
Enter numbers: 34
Enter numbers: 2
Enter numbers: 5
Enter numbers: 24
Enter numbers: 
Enter a number or the number is already existing
Enter numbers: 454
Enter numbers: 34
Enter numbers: 43
Enter numbers: 34
Enter numbers: done
[34.0, 34.0, 2.0, 5.0, 24.0, 454.0, 34.0, 43.0, 34.0]
The maximum number is:  454.0
The minimum number is:  2.0


## Dictionaries

In [7]:
eng2sp = dict()
eng2sp

{}

In [8]:
eng2sp['one'] = 'uno'
eng2sp

{'one': 'uno'}

In [9]:
eng2sp = {'one':'uno','two':'dos','three':'tres'}
eng2sp

{'one': 'uno', 'two': 'dos', 'three': 'tres'}

In [10]:
# order of items is unpredictable
print(eng2sp['two'])

dos


In [11]:
print(eng2sp['fsd'])

KeyError: 'fsd'

In [12]:
# return number of key value pairs

print(len(eng2sp))

3


In [13]:
# in operator checks existance of key in the item
'one' in eng2sp

True

In [14]:
'ten' in eng2sp

False

In [15]:
# to check value in item

vals = list(eng2sp.values())
vals

['uno', 'dos', 'tres']

In [16]:
'uno' in vals

True

* in operator may take more time for searching list and same time for dictionary(hash table)

### Exercise 1:
Write a program that reads words in words.txt and stores them as keys in a dictionary. It doesnt matter what the values are. Then you can use the in operator to check whether a string is in the dictionary.

In [18]:
fhand = open('words.txt')

word_dict = dict()

for line in fhand:
    words = line.split()
    for word in words:
        if word not in word_dict.keys():
            word_dict[word] = ''
print(word_dict)


{'Writing': '', 'programs': '', 'or': '', 'programming': '', 'is': '', 'a': '', 'very': '', 'creative': '', 'and': '', 'rewarding': '', 'activity': '', 'You': '', 'can': '', 'write': '', 'for': '', 'many': '', 'reasons': '', 'ranging': '', 'from': '', 'making': '', 'your': '', 'living': '', 'to': '', 'solving': '', 'difficult': '', 'data': '', 'analysis': '', 'problem': '', 'having': '', 'fun': '', 'helping': '', 'someone': '', 'else': '', 'solve': '', 'This': '', 'book': '', 'assumes': '', 'that': '', '{\\em': '', 'everyone}': '', 'needs': '', 'know': '', 'how': '', 'program': '', 'once': '', 'you': '', 'program,': '', 'will': '', 'figure': '', 'out': '', 'what': '', 'want': '', 'do': '', 'with': '', 'newfound': '', 'skills': '', 'We': '', 'are': '', 'surrounded': '', 'in': '', 'our': '', 'daily': '', 'lives': '', 'computers': '', 'laptops': '', 'cell': '', 'phones': '', 'think': '', 'of': '', 'these': '', 'as': '', 'personal': '', 'assistants': '', 'who': '', 'take': '', 'care': '', 

In [21]:
'needs' in word_dict

True

### Dictionary as a set of counters

You could create a dictionary with characters as keys and counters as the corresponding values. The first time you see a character, you would add an item to the dictionary. After that you would increment the value of an existing item. 


In [22]:
word = 'brontasaurus'
d = dict()
for c in word:
    if c not in d:
        d[c]=1
    else:
        d[c] = d[c]+1
print(d)

{'b': 1, 'r': 2, 'o': 1, 'n': 1, 't': 1, 'a': 2, 's': 2, 'u': 2}


In [23]:
# get method takes a key and a default value

counts = {'chunk':1,'annie':42,'jan':100}
counts.get('annie')

42

In [24]:
counts.get('annie3',0)

0

In [25]:
# counter using get

word = 'brontosaurus'
d = dict()
for c in word:
    d[c]=d.get(c,0) + 1
print(d)


{'b': 1, 'r': 2, 'o': 2, 'n': 1, 't': 1, 's': 2, 'a': 1, 'u': 2}


### Dictionaries and files

In [27]:
fname = input('enter the filename: ')
try:
    fhand = open(fname)
except:
    print('file cannot be opened')
    exit()
    
counts = dict()
for line in fhand:
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1
print(counts)

enter the filename: words.txt
{'Writing': 1, 'programs': 2, 'or': 1, 'programming': 1, 'is': 2, 'a': 3, 'very': 2, 'creative': 1, 'and': 5, 'rewarding': 1, 'activity': 1, 'You': 1, 'can': 4, 'write': 1, 'for': 1, 'many': 2, 'reasons': 1, 'ranging': 2, 'from': 2, 'making': 1, 'your': 2, 'living': 1, 'to': 16, 'solving': 1, 'difficult': 1, 'data': 1, 'analysis': 1, 'problem': 2, 'having': 1, 'fun': 1, 'helping': 1, 'someone': 1, 'else': 1, 'solve': 1, 'This': 1, 'book': 1, 'assumes': 1, 'that': 4, '{\\em': 1, 'everyone}': 1, 'needs': 1, 'know': 2, 'how': 2, 'program': 1, 'once': 1, 'you': 4, 'program,': 1, 'will': 1, 'figure': 1, 'out': 1, 'what': 2, 'want': 1, 'do': 5, 'with': 2, 'newfound': 1, 'skills': 1, 'We': 2, 'are': 3, 'surrounded': 1, 'in': 2, 'our': 5, 'daily': 1, 'lives': 1, 'computers': 5, 'laptops': 1, 'cell': 1, 'phones': 1, 'think': 1, 'of': 5, 'these': 1, 'as': 1, 'personal': 1, 'assistants': 1, 'who': 1, 'take': 1, 'care': 1, 'things': 3, 'on': 2, 'behalf': 2, 'The': 1, 

### Looping and dictionaries

In [28]:
counts = {'chuck':1,'annie':42,'jan':100}
counts

{'chuck': 1, 'annie': 42, 'jan': 100}

In [29]:
for key in counts:
    print(key, counts[key])

chuck 1
annie 42
jan 100


In [30]:
for key in counts:
    if counts[key]>10:
        print(key, counts[key])
    

annie 42
jan 100


In [34]:
# print dictionary in alphabetical order

counts = {'chuck':1, 'annie':42, 'jan':100}
lst = list(counts.keys())
lst.sort()
for key in lst:
    print(key,counts[key])


annie 42
chuck 1
jan 100


### Advanced text parsing


In [35]:
# lower, punctuation, translate
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [36]:
import string
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:',fname)
    exit()

counts = dict()

for line in fhand:
    #print(line)
    line = line.rstrip()
    line = line.translate(line.maketrans('','',string.punctuation))
   
    line = line.lower()
    #line = line.strip()
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1 
    #print(line)
print(counts)


Enter the file name: romeo.txt
{'but': 1, 'soft': 1, 'what': 1, 'light': 1, 'through': 1, 'yonder': 1, 'window': 1, 'breaks': 1, 'it': 1, 'is': 3, 'the': 3, 'east': 1, 'and': 3, 'juliet': 1, 'sun': 2, 'arise': 1, 'fair': 1, 'kill': 1, 'envious': 1, 'moon': 1, 'who': 1, 'already': 1, 'sick': 1, 'pale': 1, 'with': 1, 'grief': 1}


### Exercise 2:
Write a program that categorizes each mail message by
which day of the week the commit was done. To do this look for lines
that start with “From”, then look for the third word and keep a running
count of each of the days of the week. At the end of the program print
out the contents of your dictionary (order does not matter).

In [38]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File does not exist')


day_cnt = dict()

for line in fhand:
    words = line.split()
    if len(words) == 0 or not words[0] == 'From':
        continue
    # if words[2] not in day_cnt:
    #     day_cnt[words[2]] = 1
    # else:
    #     day_cnt[words[2]] += 1
    day_cnt[words[2]] = day_cnt.get(words[2], 0) + 1
print(day_cnt)

Enter the file name: mbox-short.txt
{'Sat': 1, 'Fri': 20, 'Thu': 6}


### Exercise 3: 
Write a program to read through a mail log, build a his-
togram using a dictionary to count how many messages have come from
each email address, and print the dictionary.

In [39]:
fname = input('Enter the file name: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()
mail_cnt = {}
for line in fhand:
    words = line.split()
    if len(words) == 0 or not words[0] == 'From':
        continue

    mail_cnt[words[1]] = mail_cnt.get(words[1], 0) + 1

print(mail_cnt)

Enter the file name: mbox-short.txt
{'stephen.marquard@uct.ac.za': 2, 'louis@media.berkeley.edu': 3, 'zqian@umich.edu': 4, 'rjlowe@iupui.edu': 2, 'cwen@iupui.edu': 5, 'gsilver@umich.edu': 3, 'wagnermr@iupui.edu': 1, 'antranig@caret.cam.ac.uk': 1, 'gopal.ramasammycook@gmail.com': 1, 'david.horwitz@uct.ac.za': 4, 'ray@media.berkeley.edu': 1}


### Exercise 4: 
Add code to the above program to figure out who has the
most messages in the file. After all the data has been read and the dic-
tionary has been created, look through the dictionary using a maximum
loop (see Chapter 5: Maximum and minimum loops) to find who has
the most messages and print how many messages the person has.

In [40]:
fname = input('Enter the file name: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()
mail_cnt = {}
for line in fhand:
    words = line.split()
    if len(words) == 0 or not words[0] == 'From':
        continue

    mail_cnt[words[1]] = mail_cnt.get(words[1], 0) + 1

# to find who sent the most messages

largest = None

for key, val in mail_cnt.items():
    if largest is None or val > largest[1]:
        largest = key, val
larkey, larval = largest
print('Person who sent the most messages: ', larkey)
print('Number of messages sent: ', larval)


Enter the file name: mbox.txt
Person who sent the most messages:  zqian@umich.edu
Number of messages sent:  195


### Exercise 5: 
This program records the domain name (instead of the
address) where the message was sent from instead of who the mail came
from (i.e., the whole email address). At the end of the program, print
out the contents of your dictionary.

In [43]:

fname = input('Enter the file name: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()
mail_cnt = {}
for line in fhand:
    words = line.split()
    if len(words) == 0 or not words[0] == 'From':
        continue
    domain = words[1].split('@')[-1]
    mail_cnt[domain] = mail_cnt.get(domain, 0) + 1

print(mail_cnt)

Enter the file name: mbox-short.txt
{'uct.ac.za': 6, 'media.berkeley.edu': 4, 'umich.edu': 7, 'iupui.edu': 8, 'caret.cam.ac.uk': 1, 'gmail.com': 1}


## Tuples

In [44]:
#Tuples are immutable

t = 'a','b','c','d','e'
t

('a', 'b', 'c', 'd', 'e')

In [45]:
t = ('a','b','c','d','e')
type(t)

tuple

In [46]:
# to create tuple with single element we have to include final comma
t1 = ('a',)
type(t1)

tuple

In [47]:
# another way

t = tuple()
print(t)

()


In [48]:
t = tuple('lupins')
t

('l', 'u', 'p', 'i', 'n', 's')

In [49]:
t = ('a','b','c','d','e')
print(t[0])

a


In [50]:
print(t[1:3])

('b', 'c')


In [51]:
# cannot modify element
t[0]='dummy'

TypeError: 'tuple' object does not support item assignment

In [53]:
# replace one tuple with another
t = ('A',) + t[1:]
t

('A', 'b', 'c', 'd', 'e')

### comparing tuples

In [54]:
(0,1,2) < (0,3,4)

True

In [55]:
(0,1,2000) < (0,3,4)

True

In [56]:
# sort list of words from longest to shortest

txt = 'but soft what light in yander window breaks'
words = txt.split()
t = list()
for word in words:
    t.append((len(word),word))
t.sort(reverse = True)

res = list()
for length,word in t:
    res.append(word)
print(res)

['yander', 'window', 'breaks', 'light', 'what', 'soft', 'but', 'in']


### Tuple assignment

In [57]:
m = ['have','fun']
x, y = m
x

'have'

In [58]:
y

'fun'

In [59]:
(x,y) = m
x,y

('have', 'fun')

In [60]:
# swap values of a variable
x,y = y,x
x,y

('fun', 'have')

In [61]:
# split email address into username and domain
addr = 'monty@python.org'

uname, domain = addr.split('@')
uname, domain

('monty', 'python.org')

In [62]:
addr = 'monty@python.org'

addrlst = addr.split('@')
addrlst

['monty', 'python.org']

In [63]:
a, b = ['niju', 'shanth']
a,b

('niju', 'shanth')

### Dictionary and Tuples


In [64]:
d = {'b':20,'a':1,'c':33}
d

{'b': 20, 'a': 1, 'c': 33}

In [65]:
t = list(d.items())
print(t)

[('b', 20), ('a', 1), ('c', 33)]


In [ ]:
t.sort()
t

### Multiple assignemnt with dictionaries

In [66]:
for key, val in list(d.items()):
    print(key,val)


b 20
a 1
c 33


In [68]:
# print in sorted order
l = list()
for key, val in d.items():
    l.append((val,key))
l

[(20, 'b'), (1, 'a'), (33, 'c')]

In [69]:
l.sort(reverse = True)
l

[(33, 'c'), (20, 'b'), (1, 'a')]

### The most common words



In [71]:
# print 10 most common words in the text
import string
fhand = open('romeo-full.txt')
counts = dict()
for line in fhand:
    line = line.translate(str.maketrans('','',string.punctuation))
    line = line.lower()
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1
#print(counts)

# sort the dictionary by value

lst = list()
for key,val in list(counts.items()):
    lst.append((val,key))
lst.sort(reverse = True)
for key,val in lst[:10]:
    print(key,val)

61 i
42 and
40 romeo
34 to
34 the
32 thou
32 juliet
30 that
29 my
24 thee


### Using tuples as keys in dictionary



In [73]:
number = 6
last = 'niju'
first = 'geo'
directory = {(last,first):30}
print(directory)
directory[last,first] = number
for last,first in directory:
    print(first, last, directory[last,first])
print(directory.items())

{('niju', 'geo'): 30}
geo niju 6
dict_items([(('niju', 'geo'), 6)])


### Sequences: strings, lists, and tuples - oh My!
#### Debugging


#### Exercise 1: 
Revise a previous program as follows: Read and parse the
“From” lines and pull out the addresses from the line. Count the num-
ber of messages from each person using a dictionary.
After all the data has been read, print the person with the most commits
by creating a list of (count, email) tuples from the dictionary. Then
sort the list in reverse order and print out the person who has the most
commits.

In [74]:
fname = input('Enter the file name: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()
mail_cnt = {}
for line in fhand:
    words = line.split()
    if len(words) == 0 or not words[0] == 'From':
        continue

    mail_cnt[words[1]] = mail_cnt.get(words[1], 0) + 1


cnt_lst = []

for key, val in list(mail_cnt.items()):
    cnt_lst.append((val, key))

cnt_lst.sort(reverse=True)

for key, val in cnt_lst:
    print(val, key)
    break



Enter the file name: mbox-short.txt
cwen@iupui.edu 5


#### Exercise 2: 
This program counts the distribution of the hour of the day
for each of the messages. You can pull the hour from the “From” line
by finding the time string and then splitting that string into parts using
the colon character. Once you have accumulated the counts for each
hour, print out the counts, one per line, sorted by hour as shown below

In [1]:
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()

mail_cnt = dict()

for line in fhand:
    words = line.split()
    if len(words) == 0 or not words[0] == 'From':
        continue
    msghrs = words[5].split(':')[0]
    mail_cnt[msghrs] = mail_cnt.get(msghrs, 0) + 1
#print(mail_cnt)

mail_cnt_lst = list(mail_cnt.items())
mail_cnt_lst.sort()
#print(mail_cnt_lst)

for key, value in mail_cnt_lst:
    print(key, value)


Enter the file name: mbox-short.txt
04 3
06 1
07 1
09 2
10 3
11 6
14 1
15 2
16 4
17 2
18 1
19 1


#### Exercise 3: 
Write a program that reads a file and prints the letters
in decreasing order of frequency. Your program should convert all the
input to lower case and only count the letters a-z. Your program should
not count spaces, digits, punctuation, or anything other than the letters
a-z. Find text samples from several different languages and see how
letter frequency varies between languages. Compare your results with
the tables at https://wikipedia.org/wiki/Letter_frequencies.

In [2]:
import string

fname = input('Enter the file name: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit()

word_cnt = {}

for line in fhand:
    line = line.translate(str.maketrans('', '', string.punctuation))
    line = line.lower()
    words = line.split()
    for word in words:
        word_cnt[word] = word_cnt.get(word, 0) + 1

word_cnt = list(word_cnt.items())
word_srt_lst = []

for key, value in word_cnt:
    word_srt_lst.append((value, key))

word_srt_lst.sort(reverse=True)

for key, value in word_srt_lst:
    print(value, key)

Enter the file name: words.txt
to 16
we 7
the 7
our 6
you 5
of 5
do 5
computers 5
and 5
that 4
can 4
what 3
things 3
are 3
a 3
your 2
would 2
with 2
very 2
us 2
this 2
ranging 2
programs 2
program 2
problem 2
on 2
next 2
many 2
like 2
language 2
know 2
knew 2
kinds 2
is 2
in 2
if 2
how 2
from 2
could 2
computer 2
behalf 2
writing 1
write 1
will 1
who 1
were 1
want 1
vasts 1
think 1
these 1
tell 1
tasks 1
take 1
surrounded 1
speak 1
someone 1
solving 1
solve 1
skills 1
rewarding 1
reptitive 1
reasons 1
question 1
programming 1
phones 1
personal 1
out 1
or 1
only 1
once 1
often 1
newfound 1
needs 1
mindnumbing 1
memory 1
me 1
making 1
living 1
lives 1
laptops 1
it 1
interestingly 1
humans 1
helping 1
helpful 1
having 1
have 1
hardware 1
fun 1
for 1
find 1
figure 1
fast 1
explain 1
everyone 1
essentially 1
em 1
else 1
difficult 1
data 1
daily 1
currentday 1
creative 1
continuously 1
cell 1
care 1
built 1
boring 1
book 1
best 1
be 1
assumes 1
assistants 1
ask 1
as 1
analysis 1
amounts 1
ac

### References

https://www.programiz.com/python-programming/methods/string/maketrans

## Chapter 11 Regular Expressions


In [6]:
str.maketrans('abl','csd','ab')

{97: None, 98: None, 108: 100}

In [8]:
# search function

import re
    
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('From:',line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [9]:
# search for lines that start with from

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.strip()
    if re.search('^From:',line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


## character matching in regular expressions


In [10]:
# search for lines that starts with 'F', followed by 2 characters followed by 'm'

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.strip()
    if re.search('^F..m:',line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [11]:
# search for lines that start with from and have an at sign

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('^From:.+@',line):
        print(line)
    

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


### Extracting data using regular expressions

In [13]:
# find email address from line

import re
s = 'A message from csev@umich.edu to cwen@iupi.edu about meeting @2pm'
lst = re.findall('\S+@\S+',s) #returns a list of found objects
print(lst)

['csev@umich.edu', 'cwen@iupi.edu']


In [14]:
# search for lines that have an at sign between characters
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('\S+@\S+',line)
    if len(x)>0:
        print(x)
    

['stephen.marquard@uct.ac.za']
['<postmaster@collab.sakaiproject.org>']
['<200801051412.m05ECIaH010327@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042308.m04N8v6O008125@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042109.m04L92hb007923@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject

In [15]:
# search for lines that have an at sign between characters
# the characters must be a letter or number

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('[a-zA-Z0-9]\S+@\S+[a-xA-Z]',line)
    if len(x)>0:
        print(x)

['stephen.marquard@uct.ac.za']
['postmaster@collab.sakaiproject.org']
['200801051412.m05ECIaH010327@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['postmaster@collab.sakaiproject.org']
['200801042308.m04N8v6O008125@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['postmaster@collab.sakaiproject.org']
['200801042109.m04L92hb007923@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject

### Combining, searching, extracting

In [18]:
# find numbers on line that starts with X

# search for lines that starts with'X' followed by any non white space character and ':'
# followed by space and any number
# the number can include decimal

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
#     if re.search('^X\S*: [0-9.]+',line):
#         print(line)
    x = re.findall('^X\S*: ([0-9.]+)',line)
    if len(x)>0:
        print(x)

['0.8475']
['0.0000']
['0.6178']
['0.0000']
['0.6961']
['0.0000']
['0.7565']
['0.0000']
['0.7626']
['0.0000']
['0.7556']
['0.0000']
['0.7002']
['0.0000']
['0.7615']
['0.0000']
['0.7601']
['0.0000']
['0.7605']
['0.0000']
['0.6959']
['0.0000']
['0.7606']
['0.0000']
['0.7559']
['0.0000']
['0.7605']
['0.0000']
['0.6932']
['0.0000']
['0.7558']
['0.0000']
['0.6526']
['0.0000']
['0.6948']
['0.0000']
['0.6528']
['0.0000']
['0.7002']
['0.0000']
['0.7554']
['0.0000']
['0.6956']
['0.0000']
['0.6959']
['0.0000']
['0.7556']
['0.0000']
['0.9846']
['0.0000']
['0.8509']
['0.0000']
['0.9907']
['0.0000']


In [23]:
# search for lines that start with 'Details: rev='
# followed by numbers and '.'
# Then print the numbe if it is greater than zero

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('^Details:.*rev=([0-9.]+)',line)
    if len(x)>0:
        print(x)

['39772']
['39771']
['39770']
['39769']
['39766']
['39765']
['39764']
['39763']
['39762']
['39761']
['39760']
['39759']
['39758']
['39757']
['39756']
['39755']
['39754']
['39753']
['39752']
['39751']
['39750']
['39749']
['39746']
['39745']
['39744']
['39743']
['39742']


In [21]:
# search for ines that start with from and a character
# followed by a two digit number between 00 and 99 followed by ':'
# then print the number if it is greater than zero

import re
hand = open('mbox.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('^From .* ([0-9][0-9]):',line)
    if len(x)>0:
        print(x)

['09']
['18']
['16']
['15']
['15']
['14']
['11']
['11']
['11']
['11']
['11']
['11']
['10']
['10']
['10']
['09']
['07']
['06']
['04']
['04']
['04']
['19']
['17']
['17']
['16']
['16']
['16']
['15']
['15']
['15']
['15']
['15']
['14']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['12']
['12']
['12']
['12']
['11']
['11']
['11']
['10']
['10']
['10']
['10']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['09']
['08']
['18']
['18']
['17']
['17']
['17']
['16']
['16']
['15']
['15']
['15']
['14']
['12']
['12']
['11']
['11']
['09']
['09']
['09']
['09']
['09']
['09']
['08']
['08']
['08']
['08']
['07']
['05']
['04']
['04']
['03']
['03']
['03']
['03']
['02']
['20']
['07']
['16']
['16']
['14']
['14']
['14']
['10']
['09']
['22']
['16']
['16']
['13']
['12']
['10']
['10']
['21']
['14']
['10']
['10']
['23']
['16']
['14']
['12']
['12']
['11']
['21']
['21']
['17']
['17']
['17']
['17']
['17']
['13']
['12']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['08']
['21']
['16']
['11']

['13']
['12']
['12']
['11']
['11']
['11']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['08']
['08']
['08']
['08']
['08']
['07']
['07']
['07']
['06']
['06']
['06']
['16']
['15']
['15']
['15']
['15']
['15']
['11']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['09']
['09']
['09']
['08']
['08']
['08']
['08']
['08']
['07']
['07']
['07']
['07']
['07']
['07']
['07']
['06']
['06']
['06']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['14']
['14']
['14']
['14']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['12']
['12']
['12']
['12']
['12']
['12']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['10']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['04']
['22']
['22']
['20']
['19']
['18']
['17']
['17']
['17']
['17']
['17']
['16']
['16']
['16']
['16']
['16']

### Escape character



In [25]:
import re
x = 'We just received $10.00 for cookies.'
y = re.findall('\$[0-9.]+',x) # inside square bracket, characters are not special, '\' --> Escape character for dollar'
#y = re.findall('[$0-9.]+',x) # inside square bracket, characters are not special
print(y)

['$10.00']


* ^ matches the beginning of the line
* $ matches the end of the line
* . matches any character
* \s matches a whitespace character
* \S matches a non whitespace character
* \* applies to the immediately preceding character and indicates to match zero or more
* *? applies to the immediately preceding character and indicates to match zero or more in non greedy mode
* \+ Applies to the immediately preceding character(s) and indicates to match one or
more times.
* +? Applies to the immediately preceding character(s) and indicates to match one
    or more times in “non-greedy mode”.
* ? Applies to the immediately preceding character(s) and indicates to match zero
or one time.
* ?? Applies to the immediately preceding character(s) and indicates to match zero
or one time in “non-greedy mode”.
* [aeiou] Matches a single character as long as that character is in the specified set.
In this example, it would match “a”, “e”, “i”, “o”, or “u”, but no other characters.
* [a-z0-9] You can specify ranges of characters using the minus sign. This example
is a single character that must be a lowercase letter or a digit.
* [ˆA-Za-z] When the first character in the set notation is a caret, it inverts the
logic. This example matches a single character that is anything other than an
uppercase or lowercase letter.
* ( ) When parentheses are added to a regular expression, they are ignored for the
purpose of matching, but allow you to extract a particular subset of the matched
string rather than the whole string when using findall().
* \b Matches the empty string, but only at the start or end of a word.
* \B Matches the empty string, but not at the start or end of a word.
* \d Matches any decimal digit; equivalent to the set [0-9].
* \D Matches any non-digit character; equivalent to the set [ˆ0-9].

In [26]:
help(re.search)

Help on function search in module re:

search(pattern, string, flags=0)
    Scan through string looking for a match to the pattern, returning
    a Match object, or None if no match was found.



### Exercise 1
Write a simple program to simulate the operation of the
grep command on Unix. Ask the user to enter a regular expression and
count the number of lines that matched the regular expression

In [1]:
import re

fname = input('Enter a file name to search: ')

regex = input('Enter a regular expression: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit(0)

ln_cnt = 0

for line in fhand:
    line = line.rstrip()
    if re.search(regex, line):
        ln_cnt += 1

print(fname, ' had ', ln_cnt, ' that matched ', regex)


Enter a file name to search: mbox-short.txt
Enter a regular expression: ^Author
mbox-short.txt  had  27  that matched  ^Author


### Exercise 2: 
Write a program to look for lines of the form:
New Revision: 39772
Extract the number from each of the lines using a regular expression
and the findall() method. Compute the average of the numbers and
print out the average as an integer.

In [2]:

import re

fname = input('Enter the file name: ')

try:
    fhand = open(fname)
except:
    print('File does not exist')
    exit(0)
count = 0
sum = 0
for line in fhand:
    line = line.rstrip()
    x = re.findall('^New Revision: ([0-9.]+)', line)
    if len(x) > 0:
        if sum == 0:
            sum = float(x[0])
        else:
            sum = sum + float(x[0])
        count += 1

print(int(sum/count))

Enter the file name: mbox-short.txt
39756


## Chapter 12 
## Networked programs

### The worlds simplest web browser


In [ ]:
# request a document and display what the server sends back

import socket
mysock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
mysock.connect(('data.pr4e.org',80))
cmd = 'GET http://data.pr4e.org/romeo.txt HTTP/1.0\r\n\r\n'.encode()
mysock.send(cmd)

while True:
    data = mysock.recv(512)
    if len(data) < 1:
        break
    print(data.decode(),end = '')
mysock.close()

In [ ]:
b'Hello world'

In [ ]:
'Hello world'.encode()

### Retrieving and image over HTTP


In [ ]:
import socket
import time

HOST = 'data.pr4e.org'
PORT = 80
mysock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
mysock.connect((HOST,PORT))
mysock.sendall(b'GET http://data.pr4e.org/cover3.jpg HTTP/1.0\r\n\r\n')
count = 0
picture = b""

while True:
    data = mysock.recv(5120)
    if len(data) < 1: break
    time.sleep(0.25)
    count = count + len(data)
    print(len(data),count)
    picture = picture + data
    
mysock.close()

# Look for the end of the header (2 CRLF)

pos = picture.find(b"\r\n\r\n")
print('Header length', pos)
print(picture[:pos].decode())

#Skip past the header and save the picture data

picture = picture[pos+4:]
fhand = open("stuff.jpg", 'wb')
fhand.write(picture)
fhand.close()

                   

### Retrieving web pages with urllib

In [1]:
import urllib.request

fhand  = urllib.request.urlopen('http://data.pr4e.org/romeo.txt')
for line in fhand:
    print(line.strip())
    #print(line.decode().strip())

b'But soft what light through yonder window breaks'
b'It is the east and Juliet is the sun'
b'Arise fair sun and kill the envious moon'
b'Who is already sick and pale with grief'


In [ ]:
# read from website and compute word ffrequency

import urllib.request, urllib.parse, urllib.error

fhand = urllib.request.urlopen('http://data.pr4e.org/romeo.txt')

counts = dict()
for line in fhand:
    words = line.decode().split()
    for word in words:
        counts[word] = counts.get(word,0) + 1

print(counts)
     


### Reading binary files using urllib


In [ ]:
import urllib.request, urllib.parse, urllib.error

img = urllib.request.urlopen('http://data.pr4e.org/cover3.jpg').read()
fhand = open('cover3.jpg','wb')
fhand.write(img)
fhand.close()

In [ ]:
# reading and writing in blocks to avoid out of memory

import urllib.request, urllib.parse, urllib.error

img = urllib.request.urlopen('http://data.pr4e.org/cover3.jpg')
fhand = open('cover3.jpg','wb')
size = 0
while True:
    info = img.read(100000)
    if len(info) < 1: break
    size = size + len(info)
    fhand.write(info)

print(size,'characters copied')
fhand.close()

### Parsing the html and scraping the web

### Parsing HTML using regular expression

In [ ]:
# search the link values within URL input 

import urllib.request, urllib.parse, urllib.error
import re
import ssl

# Ignore SSL certificate errors

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = urllib.request.urlopen(url).read()
links = re.findall(b'href="(http[s]?://.*?)"',html)

for link in links:
    print(link.decode())

### Parsing HTML using beautiful soup

In [ ]:
import urllib.request, urllib.parse, urllib.error
from bs4 import BeautifulSoup
import ssl

# ignore ssl certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = urllib.request.urlopen(url,context = ctx).read()
soup = BeautifulSoup(html,'html.parser')


# Retrieve all of the anchor tags

tags = soup('a')
for tag in tags:
    print(tag.get('href',None))
    


In [ ]:
from urllib.request import urlopen
from bs4 import BeautifulSoup
import ssl
# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = urlopen(url, context=ctx).read()
soup = BeautifulSoup(html, "html.parser")

# Retrieve all of the anchor tags
tags = soup('a')

for tag in tags:
# Look at the parts of a tag
    print('TAG:', tag)
    print('URL:', tag.get('href', None))
    print('Contents:', tag.contents[0])
    print('Attrs:', tag.attrs)


### Exercise 1: 
Change the socket program socket1.py to prompt the user
for the URL so it can read any web page. You can use split('/') to
break the URL into its component parts so you can extract the host
name for the socket connect call. Add error checking using try and
except to handle the condition where the user enters an improperly
formatted or non-existent URL.
### Exercise 2: 
Change your socket program so that it counts the number
of characters it has received and stops displaying any text after it has
shown 3000 characters. The program should retrieve the entire docu-
ment and count the total number of characters and display the count
of the number of characters at the end of the document.
### Exercise 3: 
Use urllib to replicate the previous exercise of (1) retrieving
the document from a URL, (2) displaying up to 3000 characters, and
(3) counting the overall number of characters in the document. Don’t
worry about the headers for this exercise, simply show the first 3000
characters of the document contents.
### Exercise 4: 
Change the urllinks.py program to extract and count para-
graph (p) tags from the retrieved HTML document and display the
count of the paragraphs as the output of your program. Do not display
the paragraph text, only count them. Test your program on several
small web pages as well as some larger web pages.
Exercise 5: (Advanced) Change the socket program so that it only shows
data after the headers and a blank line have been received. Remember
that recv receives characters (newlines and all), not lines.

## Using web services 

### Parsing XML

In [ ]:
import xml.etree.ElementTree as ET

In [ ]:
data = '''
<person>
    <name>Chuck</name>
    <phone type="intl">
        +1 734 303 4456
    </phone>
    <email hide="yes" />
</person>'''

tree = ET.fromstring(data)
print('Name:', tree.find('name').text)
print('Attr:', tree.find('email').get('hide') )
    

### Looping through nodes


In [ ]:
import xml.etree.ElementTree as ET

input = '''
<stuff>
    <users>
        <user x="2">
            <id>001</id>
            <name>Chuck</name>
        </user>
        <user x="7">
            <id>009</id>
            <name>Brent</name>
        </user>
    </users>
</stuff>
'''

stuff = ET.fromstring(input)
lst = stuff.findall('users/user')
print('User count:',len(lst))

for item in lst:
    print('Name',item.find('name').text)
    print('Id',item.find('id').text)
    print('Attribute',item.get('x'))
    

In [ ]:
import xml.etree.ElementTree as ET
input = '''
<stuff>
    <users>
        <user x='2'>
            <id>001</id>
            <name>Chuck</name>
        </user>
        <user x='7'>
            <id>009</id>
            <name>Brent</name>
        </user>
    </users>
</stuff>
'''

stuff = ET.fromstring(input)

lst = stuff.findall('users/user')
print('User count:', len(lst))

lst2 = stuff.findall('user')
print('User count:', len(lst2))


### JavaScript Object Notation - JSON


### Parsing JSON

In [ ]:
import json
data = '''
[
{"id":"001",
"x":"2",
"name":"chuck"
},
{"id":"009",
"x":"7",
"name":"Brent"
}
]
'''

info = json.loads(data)
print('User count:',len(info))

for item in info:
    print('Name', item['name'])
    print('Id',item['id'])
    print('Attribute', item['x'])
            

### Application programming interfaces

### Application 1:
#### Google geocoding web service

In [ ]:
# call the google geocoding API, and extract information from the returned JSON

import urllib.request, urllib.parse, urllib.error
import json
import ssl

api_key = False
# If you have a google places API key, enter it here
#api_key = ''

if api_key is False:
    api_key = 42
    serviceurl = 'http://py4e-data.dr-chuck.net/json?'

else:
    serviceurl = 'https://maps.googleapis.com/maps/api/geocode/json?'
    
# Ignore SSL certificate errors

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True :
    address = input('Enter location: ')
    if len(address) < 1: break
        
    parms = dict()
    
    parms['address'] = address
    
    if api_key is not False: parms['key'] = api_key
        
    url = serviceurl + urllib.parse.urlencode(parms)
    print('Retrieving', url)
    uh = urllib.request.urlopen(url,context=ctx)
    data = uh.read().decode()
    print('Retrieved', len(data),'characters')
    
    try:
        js = json.loads(data)
    except:
        js = None
    if not js or 'status' not in js or js['status'] != 'OK':
        print('==== Failure to Retrieve ====')
        print(data)
        continue
    print(json.dumps(js,indent = 4))
    lat = js['results'][0]['geometry']['location']['lat']
    lng = js['results'][0]['geometry']['location']['lng']
    print('lat',lat, 'lng',lng)
    location = js['results'][0]['formatted_address']
    print(location)
    

### Application 2:
#### Twitter

In [ ]:
# retrieve the timeline for a particular twitter user and return it in JSON format

import urllib.request, urllib.parse, urllib.error
import twurl
import ssl

# create app and get the four strings, put them in hidden.py

TWITTER_URL = 'https://api.twitter.com/1.1/statuses/user_timeline.json'

# Ignore SSL certificate error

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    print('')
    acct = input('Enter twitter Account')
    if (len(acct) < 1): break
    url = twurl.augment(TWITTER_URL, {'screen_name': acct, 'count':'2'})
    print('Retrieving', url)
    connection = urllib.request.urlopen(url, context = ctx)
    data = connection.read().decode()
    print(data[:250])
    headers = dict(connection.getheaders())
    # print headers
    
    print('Remaining', headers['x-rate-limit-remaining'])
    

In [ ]:
# get friends information

import urllib.request, urllib.parse, urllib.error
import twurl
import json
import ssl

TWITTER_URL = 'https://api.twitter.com/1.1/friends/list.json'
# Ignore SSL certificate error

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    print('')
    acct = input('Enter Twitter Account:')
    if (len(acct)<1): break
    url = twurl.augment(TWITTER_URL, {'screen_name':acct, 'count':'5'})
    print('Retrieving', url)
    connection = urllib.request.urlopen(url, context=ctx)
    data = connection.read().decode()
    js = json.loads(data)
    print(json.dumps(js,indent=2))
    
    headers = dict(connection.getheaders())
    print('Remaining', headers['x-rate-limit-remaining'])
    
    for u in js['users']:
        print(u['screen_name'])
        if 'status' not in u:
            print('  * No status found')
            continue
        s = u['status']['text']
        print('  ', s[:50])

### Object oriented programming

In [ ]:
stuff = list()
stuff.append('python')
stuff.append('chuck')
stuff.sort()
print(stuff[0])
print(stuff.__getitem__(0))
print(list.__getitem__(stuff,0))

In [ ]:
stuff = list()
dir(stuff)

### Starting with programs

In [ ]:
# elevator conversion program

usf = input('Enter the US Floor number:')
wf = int(usf) - 1
print('Non-US Floor Number is' , wf)


In [ ]:
import urllib.request, urllib.parse, urllib.error
from bs4 import BeautifulSoup
import ssl

# Ignore SSL certificatie error

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = urllib.request.urlopen(url, context = ctx).read()
soup = BeautifulSoup(html, 'html.parser')


# retrieve all of the anchor tags

tags = soup('a')
for tag in tags:
    print(tag.get('href', None))


### Subdividing a problem



#### Our first python object

In [ ]:
class PartyAnimal:
    x = 0
    def party(self):
        self.x = self.x + 1
        print('so far',self.x)

an = PartyAnimal()
an.party()
an.party()
an.party()
PartyAnimal.party(an)

### Classes as types


In [ ]:
class PartyAnimal:
    x = 0
    
    def party(self):
        self.x = self.x + 1
        print('So far',self.x)
        
an = PartyAnimal()
print('Type', type(an))
print('Dir', dir(an))
print('Type', type(an.x))
print('Type', type(an.party))
        

### Object life cycle


In [ ]:
# add construction and destruction

class PartyAnimal:
    x = 0
    
    def __init__(self):
        print('I am constructed')
    def party(self):
        self.x = self.x + 1
        print('so far', self.x)
        
    def __del__(self):
        print('I am destructed', self.x)
    

In [ ]:
an = PartyAnimal()
an.party()
an.party()
an = 42
print('an contains an',an)

### Multiple instances

In [ ]:
class PartyAnimal:
    x = 0
    name = ''
    def __init__(self, nam):
        self.name = nam
        print(self.name, 'constructed')
        
    def party(self):
        self.x = self.x + 1
        print(self.name, 'party count', self.x)

In [ ]:
s = PartyAnimal('Sally')
j = PartyAnimal('Jim')

s.party()
j.party()
s.party()


### Inheritance

* Create a new class by extending an existing class. 
* Original class is called the parent class and the new class is called the child class.

In [ ]:
#from party import PartyAnimal

In [ ]:
class CricketFan(PartyAnimal):
    points = 0
    def six(self):
        self.points = self.points + 6
        self.party()
        
        print(self.name,'points',self.points)
s = PartyAnimal('Sally')
s.party()
j = CricketFan('Jim')
j.party()
j.six()
print(dir(j))

### Woring with database tables


In [ ]:

import sqlite3

conn = sqlite3.connect('music.sqlite')
cur = conn.cursor()

cur.execute('DROP TABLE IF EXISTS Tracks')
cur.execute('CREATE TABLE Tracks (title TEXT, plays INTEGER)')

conn.close()


In [ ]:
# insert records in the table

import sqlite3

conn = sqlite3.connect('music.sqlite')
cur = conn.cursor()

cur.execute('INSERT INTO Tracks (title,plays) VALUES (?,?)', ('Thunderstruck',20))
cur.execute('INSERT INTO Tracks (title,plays) VALUES (?,?)',('My Way', 15))

conn.commit()

print('Tracks:')
cur.execute('SELECT title, plays FROM Tracks')
for row in cur:
    print(row)
    
cur.execute('DELETE FROM Tracks WHERE plays < 100')
conn.commit()
cur.close()

### Spidering twitter using a database



In [ ]:
from urllib.request import urlopen
import urllib.error
import twurl
import json
import sqlite3
import ssl

TWITTER_URL = 'https://api.twitter.com/1.1/friends/list.json'

conn = sqlite3.connect('spider.sqlite')
cur = conn.cursor()

cur.execute('''
            CREATE TABLE IF NOT EXISTS Twitter
            (name TEXT, retrieved INTEGER, friends INTEGER)''')

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    acct = input('Enter a Twitter account, or quit: ')
    if (acct == 'quit'): break
    if (len(acct) < 1):
        cur.execute('SELECT name FROM Twitter WHERE retrieved = 0 LIMIT 1')
        try:
            acct = cur.fetchone()[0]
        except:
            print('No unretrieved Twitter accounts found')
            continue

    url = twurl.augment(TWITTER_URL, {'screen_name': acct, 'count': '5'})
    print('Retrieving', url)
    connection = urlopen(url, context=ctx)
    data = connection.read().decode()
    headers = dict(connection.getheaders())

    print('Remaining', headers['x-rate-limit-remaining'])
    js = json.loads(data)
    # Debugging
    # print json.dumps(js, indent=4)

    cur.execute('UPDATE Twitter SET retrieved=1 WHERE name = ?', (acct, ))

    countnew = 0
    countold = 0
    for u in js['users']:
        friend = u['screen_name']
        print(friend)
        cur.execute('SELECT friends FROM Twitter WHERE name = ? LIMIT 1',
                    (friend, ))
        try:
            count = cur.fetchone()[0]
            cur.execute('UPDATE Twitter SET friends = ? WHERE name = ?',
                        (count+1, friend))
            countold = countold + 1
        except:
            cur.execute('''INSERT INTO Twitter (name, retrieved, friends)
                        VALUES (?, 0, 1)''', (friend, ))
            countnew = countnew + 1
    print('New accounts=', countnew, ' revisited=', countold)
    conn.commit()

cur.close()

In [ ]:
import sqlite3

conn = sqlite3.connect('spider.sqlite')
cur = conn.cursor()
cur.execute('SELECT * FROM Twitter')
count = 0
for row in cur:
    print(row)
    count = count + 1
print(count, 'rows.')
cur.close()

### Basic data modelling

The act of deciding how to breakup your application data into multiple tables and establishing the relationships between the tables is called data modeling.

In [ ]:
import urllib.request, urllib.parse, urllib.error
import twurl
import json
import sqlite3
import ssl

TWITTER_URL = 'https://api.twitter.com/1.1/friends/list.json'

conn = sqlite3.connect('friends.sqlite')
cur = conn.cursor()

cur.execute('''CREATE TABLE IF NOT EXISTS People
            (id INTEGER PRIMARY KEY, name TEXT UNIQUE, retrieved INTEGER)''')
cur.execute('''CREATE TABLE IF NOT EXISTS Follows
            (from_id INTEGER, to_id INTEGER, UNIQUE(from_id, to_id))''')

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    acct = input('Enter a Twitter account, or quit: ')
    if (acct == 'quit'): break
    if (len(acct) < 1):
        cur.execute('SELECT id, name FROM People WHERE retrieved=0 LIMIT 1')
        try:
            (id, acct) = cur.fetchone()
        except:
            print('No unretrieved Twitter accounts found')
            continue
    else:
        cur.execute('SELECT id FROM People WHERE name = ? LIMIT 1',
                    (acct, ))
        try:
            id = cur.fetchone()[0]
        except:
            cur.execute('''INSERT OR IGNORE INTO People
                        (name, retrieved) VALUES (?, 0)''', (acct, ))
            conn.commit()
            if cur.rowcount != 1:
                print('Error inserting account:', acct)
                continue
            id = cur.lastrowid

    url = twurl.augment(TWITTER_URL, {'screen_name': acct, 'count': '100'})
    print('Retrieving account', acct)
    try:
        connection = urllib.request.urlopen(url, context=ctx)
    except Exception as err:
        print('Failed to Retrieve', err)
        break

    data = connection.read().decode()
    headers = dict(connection.getheaders())

    print('Remaining', headers['x-rate-limit-remaining'])

    try:
        js = json.loads(data)
    except:
        print('Unable to parse json')
        print(data)
        break

    # Debugging
    # print(json.dumps(js, indent=4))

    if 'users' not in js:
        print('Incorrect JSON received')
        print(json.dumps(js, indent=4))
        continue

    cur.execute('UPDATE People SET retrieved=1 WHERE name = ?', (acct, ))

    countnew = 0
    countold = 0
    for u in js['users']:
        friend = u['screen_name']
        print(friend)
        cur.execute('SELECT id FROM People WHERE name = ? LIMIT 1',
                    (friend, ))
        try:
            friend_id = cur.fetchone()[0]
            countold = countold + 1
        except:
            cur.execute('''INSERT OR IGNORE INTO People (name, retrieved)
                        VALUES (?, 0)''', (friend, ))
            conn.commit()
            if cur.rowcount != 1:
                print('Error inserting account:', friend)
                continue
            friend_id = cur.lastrowid
            countnew = countnew + 1
        cur.execute('''INSERT OR IGNORE INTO Follows (from_id, to_id)
                    VALUES (?, ?)''', (id, friend_id))
    print('New accounts=', countnew, ' revisited=', countold)
    print('Remaining', headers['x-rate-limit-remaining'])
    conn.commit()
cur.close()

In [ ]:
### Using joins to retrieve data

import sqlite3

conn = sqlite3.connect('friends.sqlite')
cur = conn.cursor()

cur.execute('SELECT * FROM People')
count = 0
print('People:')
for row in cur:
    if count < 5: print(row)
    count = count + 1
print(count, 'rows.')

cur.execute('SELECT * FROM Follows')
count = 0
print('Follows:')
for row in cur:
    if count < 5: print(row)
    count = count + 1
print(count, 'rows.')

cur.execute('''SELECT * FROM Follows JOIN People
            ON Follows.to_id = People.id
            WHERE Follows.from_id = 2''')
count = 0
print('Connections for id=2:')
for row in cur:
    if count < 5: print(row)
    count = count + 1
print(count, 'rows.')

cur.close()


In [9]:
lsttt = iter(['niju','shanth','goo'])
print(type(lsttt
        ))


<class 'list_iterator'>


In [13]:
next(lsttt)

StopIteration: 

In [6]:
next(iter(lsttt))

'niju'

TypeError: 'list' object is not an iterator